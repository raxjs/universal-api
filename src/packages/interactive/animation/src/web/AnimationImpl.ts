import { Animation, AnimationAction, AnimationActionAnimate, AnimationData, AnimationOptions } from '../types';
import { getMergedOptions } from '../common';
import { CONTAINER_NAME } from '@utils/constant';

const ANIMATION_RUNNING = '__UNI_API_ANIMATION_RUNNING__';
const ANIMATION_CANCEL = '__UNI_API_ANIMATION_CANCEL__';
const ANIMATION_ORIGINAL_STYLE = '__UNI_API_ANIMATION_ORIGINAL_STYLE__';

/**
 * normalize unit
 * @param value
 * @param unit
 * @param force
 */
function normalizeUnit(value: number | string, unit: string, force = false): string {
  if (force || typeof value === 'number') {
    return `${value}${unit}`;
  }
  return value;
}

/**
 * handle animation actions queue
 * @param actions
 * @param fn
 * @param onFinish
 */
function handleActionsQueue(
  actions: AnimationAction[],
  fn: (action: AnimationAction, callback: () => void) => any,
  onFinish?: () => any,
) {
  actions = actions.slice();
  const action = actions.shift();
  if (action) {
    fn(action, () => handleActionsQueue(actions, fn, onFinish));
  } else if (onFinish) {
    onFinish();
  }
}


function applyWebAnimation(actions: AnimationAction[], dom?: HTMLElement) {
  // If `dom` is not HTML Node, ignore
  if (!(dom && dom.nodeType === 1)) {
    return;
  }

  // clear the last animation
  if (dom[ANIMATION_RUNNING]) {
    (dom[ANIMATION_CANCEL] || []).forEach((t) => clearTimeout(t));
  } else {
    // cache original style
    dom[ANIMATION_ORIGINAL_STYLE] = {
      transitionProperty: dom.style.transitionProperty,
      transitionDuration: dom.style.transitionDuration,
      transitionDelay: dom.style.transitionDelay,
      transitionTimingFunction: dom.style.transitionTimingFunction,
      transformOrigin: dom.style.transformOrigin,
    };
  }
  dom[ANIMATION_RUNNING] = true;
  dom[ANIMATION_CANCEL] = [];

  /**
   * Apply animation action, return the current cost time
   * @param action
   */
  const applyAction = (action): number => {
    const { transition, transformOrigin } = action.option;

    let transform = '';
    action.animates.forEach((animate) => {
      if (animate.type === 'style') {
        const [property, value] = animate.args;
        dom.style[property] = value;
      } else {
        transform += ` ${animate.type}(${animate.args.join(',')})`;
      }
    });
    if (transform) {
      dom.style.transform = transform;
    }

    dom.style.transitionProperty = 'all';
    dom.style.transitionDuration = `${transition.duration}ms`;
    dom.style.transitionDelay = `${transition.delay}ms`;
    dom.style.transitionTimingFunction = transition.timingFunction;
    dom.style.transformOrigin = transformOrigin;

    const cost = transition.delay + transition.duration;
    return isNaN(cost) ? 0 : cost;
  };

  const saveCancel = (timer: any) => {
    dom[ANIMATION_CANCEL] = dom[ANIMATION_CANCEL] || [];
    dom[ANIMATION_CANCEL].push(timer);
  };

  let delay = 0;
  handleActionsQueue(actions, (action, callback) => {
    saveCancel(
      setTimeout(() => {
        delay = applyAction(action);
        callback();
      }, delay),
    );
  }, () => {
    saveCancel(
      // Restore the style at the end of the animation
      setTimeout(() => {
        Object.assign(dom.style, dom[ANIMATION_ORIGINAL_STYLE] || {});
        dom[ANIMATION_RUNNING] = null;
        dom[ANIMATION_CANCEL] = null;
        dom[ANIMATION_ORIGINAL_STYLE] = null;
      }, delay),
    );
  });
}


export default class AnimationImpl implements Animation {
  private options: AnimationOptions;
  private actions: AnimationAction[];
  private currentTransform: Record<string, AnimationActionAnimate>;
  private currentStepAnimates: AnimationActionAnimate[];

  constructor(options?: AnimationOptions) {
    this.options = getMergedOptions(CONTAINER_NAME.WEB, options);
    this.actions = [];
    this.currentTransform = {};
    this.currentStepAnimates = [];
  }

  export(dom?: HTMLElement): AnimationData {
    const actions = this.actions.slice();
    applyWebAnimation(actions, dom);

    this.actions = [];
    return {
      actions,
    };
  }

  step(options?: AnimationOptions): Animation {
    options = options ? getMergedOptions(CONTAINER_NAME.WEB, options) : this.options;

    this.currentStepAnimates.forEach((animate) => {
      let key = animate.type;
      if (key === 'style') {
        key += `.${animate.args[0]}`;
      }
      this.currentTransform[key] = animate;
    });

    const animates = [];
    Object.keys(this.currentTransform).forEach((key) => {
      animates.push(this.currentTransform[key]);
    });

    this.actions.push({
      animates,
      option: {
        transformOrigin: options.transformOrigin,
        transition: {
          delay: options.delay,
          duration: options.duration,
          timingFunction: options.timingFunction,
        },
      },
    });
    this.currentStepAnimates = [];

    return this;
  }

  opacity(value: number): Animation {
    this.currentStepAnimates.push({
      type: 'style',
      args: ['opacity', value],
    });
    return this;
  }

  backgroundColor(value: string): Animation {
    this.currentStepAnimates.push({
      type: 'style',
      args: ['background-color', value],
    });
    return this;
  }

  width(value: number | string): Animation {
    this.currentStepAnimates.push({
      type: 'style',
      args: ['width', normalizeUnit(value, 'px')],
    });
    return this;
  }

  height(value: number | string): Animation {
    this.currentStepAnimates.push({
      type: 'style',
      args: ['height', normalizeUnit(value, 'px')],
    });
    return this;
  }

  left(value: number | string): Animation {
    this.currentStepAnimates.push({
      type: 'style',
      args: ['left', normalizeUnit(value, 'px')],
    });
    return this;
  }

  right(value: number | string): Animation {
    this.currentStepAnimates.push({
      type: 'style',
      args: ['right', normalizeUnit(value, 'px')],
    });
    return this;
  }

  top(value: number | string): Animation {
    this.currentStepAnimates.push({
      type: 'style',
      args: ['top', normalizeUnit(value, 'px')],
    });
    return this;
  }

  bottom(value: number | string): Animation {
    this.currentStepAnimates.push({
      type: 'style',
      args: ['bottom', normalizeUnit(value, 'px')],
    });
    return this;
  }

  rotate(angle = 0): Animation {
    this.currentStepAnimates.push({
      type: 'rotate',
      args: [normalizeUnit(angle, 'deg', true)],
    });
    return this;
  }

  rotate3d(x = 0, y = 0, z = 0, angle = 0): Animation {
    this.currentStepAnimates.push({
      type: 'rotate3d',
      args: [x, y, z, normalizeUnit(angle, 'deg', true)],
    });
    return this;
  }

  rotateX(angle = 0): Animation {
    this.currentStepAnimates.push({
      type: 'rotateX',
      args: [normalizeUnit(angle, 'deg', true)],
    });
    return this;
  }

  rotateY(angle = 0): Animation {
    this.currentStepAnimates.push({
      type: 'rotateY',
      args: [normalizeUnit(angle, 'deg', true)],
    });
    return this;
  }

  rotateZ(angle = 0): Animation {
    this.currentStepAnimates.push({
      type: 'rotateZ',
      args: [normalizeUnit(angle, 'deg', true)],
    });
    return this;
  }

  scale(sx: number, sy: number): Animation {
    if (sx === undefined) sx = 1;
    if (sy === undefined) sy = sx;

    this.currentStepAnimates.push({
      type: 'scale',
      args: [sx, sy],
    });
    return this;
  }

  scale3d(sx = 1, sy = 1, sz = 1): Animation {
    this.currentStepAnimates.push({
      type: 'scale3d',
      args: [sx, sy, sz],
    });
    return this;
  }

  scaleX(scale = 1): Animation {
    this.currentStepAnimates.push({
      type: 'scaleX',
      args: [scale],
    });
    return this;
  }

  scaleY(scale = 1): Animation {
    this.currentStepAnimates.push({
      type: 'scaleY',
      args: [scale],
    });
    return this;
  }

  scaleZ(scale = 1): Animation {
    this.currentStepAnimates.push({
      type: 'scaleZ',
      args: [scale],
    });
    return this;
  }

  translate(tx = 0, ty = 0): Animation {
    this.currentStepAnimates.push({
      type: 'translate',
      args: [normalizeUnit(tx, 'px', true), normalizeUnit(ty, 'px', true)],
    });
    return this;
  }

  translate3d(tx = 0, ty = 0, tz = 0): Animation {
    this.currentStepAnimates.push({
      type: 'translate3d',
      args: [
        normalizeUnit(tx, 'px', true),
        normalizeUnit(ty, 'px', true),
        normalizeUnit(tz, 'px', true),
      ],
    });
    return this;
  }

  translateX(translation = 0): Animation {
    this.currentStepAnimates.push({
      type: 'translateX',
      args: [normalizeUnit(translation, 'px', true)],
    });
    return this;
  }

  translateY(translation = 0): Animation {
    this.currentStepAnimates.push({
      type: 'translateY',
      args: [normalizeUnit(translation, 'px', true)],
    });
    return this;
  }

  translateZ(translation = 0): Animation {
    this.currentStepAnimates.push({
      type: 'translateZ',
      args: [normalizeUnit(translation, 'px', true)],
    });
    return this;
  }

  skew(ax = 0, ay = 0): Animation {
    this.currentStepAnimates.push({
      type: 'skew',
      args: [normalizeUnit(ax, 'deg', true), normalizeUnit(ay, 'deg', true)],
    });
    return this;
  }

  skewX(angle = 0): Animation {
    this.currentStepAnimates.push({
      type: 'skewX',
      args: [normalizeUnit(angle, 'deg', true)],
    });
    return this;
  }

  skewY(angle = 0): Animation {
    this.currentStepAnimates.push({
      type: 'skewY',
      args: [normalizeUnit(angle, 'deg', true)],
    });
    return this;
  }

  matrix(a = 1, b = 0, c = 0, d = 1, tx = 0, ty = 0): Animation {
    this.currentStepAnimates.push({
      type: 'matrix',
      args: [a, b, c, d, tx, ty],
    });
    return this;
  }

  matrix3d(
    a1 = 1, b1 = 0, c1 = 0, d1 = 0,
    a2 = 0, b2 = 1, c2 = 0, d2 = 0,
    a3 = 0, b3 = 0, c3 = 1, d3 = 0,
    a4 = 0, b4 = 0, c4 = 0, d4 = 1,
  ): Animation {
    this.currentStepAnimates.push({
      type: 'matrix3d',
      args: [
        a1, b1, c1, d1,
        a2, b2, c2, d2,
        a3, b3, c3, d3,
        a4, b4, c4, d4,
      ],
    });
    return this;
  }
}

