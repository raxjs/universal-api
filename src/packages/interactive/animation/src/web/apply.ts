import { AnimationData } from '../types';

export default function applyWebAnimation(data: AnimationData, dom?: HTMLElement) {
  // If `dom` is not HTML Node, ignore
  if (!(dom && dom.nodeType === 1)) {
    return;
  }

  /**
   * Apply animation action, return the current cost time
   * @param action
   */
  const applyAction = (action): number => {
    const { transition, transformOrigin } = action.option;

    let transform = '';
    action.animates.forEach((animate) => {
      if (animate.type === 'style') {
        const [property, value] = animate.args;
        dom.style[property] = value;
      } else {
        transform += ` ${animate.type}(${animate.args.join(',')})`;
      }
    });
    if (transform) {
      dom.style.transform = transform;
    }

    dom.style.transitionProperty = 'all';
    dom.style.transitionDuration = `${transition.duration}ms`;
    dom.style.transitionDelay = `${transition.delay}ms`;
    dom.style.transitionTimingFunction = transition.timingFunction;
    dom.style.transformOrigin = transformOrigin;

    const cost = transition.delay + transition.duration;
    return isNaN(cost) ? 0 : cost;
  };

  let delay = 0;
  data.actions.forEach((action) => {
    if (delay > 0) {
      setTimeout(() => {
        delay += applyAction(action);
      }, delay);
    } else {
      delay += applyAction(action);
    }
  });
}
