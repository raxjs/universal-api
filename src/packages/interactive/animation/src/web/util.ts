import { AnimationAction, AnimationOptions } from '../types';
import { normalizeOptions } from '../common';
import { CONTAINER_NAME } from '@utils/constant';

/**
 * normalize unit
 * @param value
 * @param unit
 * @param force
 */
export function normalizeUnit(value: number | string, unit: string, force = false): string {
  if (force || typeof value === 'number') {
    return `${value}${unit}`;
  }
  return value;
}

/**
 * merge default options
 * @param options
 */
export function getDefaultOptions(options?: AnimationOptions): AnimationOptions {
  return normalizeOptions(
    {
      duration: 400,
      timingFunction: 'linear',
      delay: 0,
      transformOrigin: '50% 50% 0',
      ...options,
    },
    CONTAINER_NAME.WEB,
  );
}

/**
 * handle animation actions queue
 * @param actions
 * @param fn
 * @param onFinish
 */
export function handleActionsQueue(
  actions: AnimationAction[],
  fn: (action: AnimationAction, callback: () => void) => any,
  onFinish?: () => any,
) {
  actions = actions.slice();
  const action = actions.shift();
  if (action) {
    fn(action, () => handleActionsQueue(actions, fn, onFinish));
  } else if (onFinish) {
    onFinish();
  }
}
