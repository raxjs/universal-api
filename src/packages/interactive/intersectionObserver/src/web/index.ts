import { CreateIntersectionObserver, OPTIONS, RECT, CallbackOptions } from '../types';

function buildMarginStr(margins: RECT): string {
  return `${Number(margins.top) || 0 }px ${ Number(margins.right) || 0 }px ${ Number(margins.bottom) || 0 }px ${ Number(margins.left) || 0 }px`;
}

function IntersectionObservers(options?: OPTIONS) {
  this.thresholds = (options && options.thresholds) ? options.thresholds : [0];
  this.initialRatio = (options && options.initialRatio) ? options.initialRatio : 0;
  this.selectAll = (options && options.selectAll) ? options.selectAll : false;
  this.callBack = () => {};
  this.observers = [];
}

function relative(root: Element | null, self, margins?: RECT) {
  const options = {
    root,
    rootMargin: margins ? buildMarginStr(margins) : '0px',
    threshold: self.thresholds,
  };
  self.observers.push(new IntersectionObserver((...args) => {
    self.callBack(...args);
  }, options));
}

IntersectionObservers.prototype.relativeTo = function (selector: string, margins?: RECT) {
  relative(document.querySelector(selector), this, margins);
  return this;
};

IntersectionObservers.prototype.relativeToViewport = function (margins?: RECT) {
  relative(null, this, margins);
  return this;
};

IntersectionObservers.prototype.observe = function (targetSelector: string, callback: (res: CallbackOptions) => any) {
  if (this.selectAll) {
    const targets = document.querySelectorAll(targetSelector);
    targets.forEach((target) => this.observers.forEach((observer) => observer.observe(target)));
  } else {
    const target = document.querySelector(targetSelector);
    this.observers.forEach((observer) => observer.observe(target));
  }
  this.callBack = (entries, observer) => {
    entries.forEach((entry) => {
      callback && typeof callback === 'function' && callback({
        boundingClientRect: entry.boundingClientRect,
        intersectionRatio: entry.intersectionRatio,
        intersectionRect: entry.intersectionRect,
        relativeRect: entry.rootBounds,
        time: entry.time,
      });
    });
  };
  return this;
};

IntersectionObservers.prototype.disconnect = function (targetSelector: string, callback: (res: CallbackOptions) => any) {
  this.callBack = () => {};
  this.observers = [];
  this.observers.forEach((observer) => {
    observer.disconnect();
  });
  return this;
};

const createIntersectionObserver: CreateIntersectionObserver = (options?, component?) => {
  return new IntersectionObservers(options);
};

export default createIntersectionObserver;
