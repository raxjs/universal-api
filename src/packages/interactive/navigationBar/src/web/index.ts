import setNavigationBarTitle from './setNavigationBarTitle';

export {
  setNavigationBarTitle,
};
export default {
  setNavigationBarTitle,
};
