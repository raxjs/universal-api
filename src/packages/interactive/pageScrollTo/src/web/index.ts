import { Options } from '../types';
import { normalize } from '../common';
import { CONTAINER_NAME } from '@utils/constant';

export default normalize((options: Options) => {
  const { selector, duration, success, fail, complete } = options;
  let { scrollTop } = options;
  const rootElement = document.documentElement;
  let top = rootElement.scrollTop;

  if (scrollTop == null) {
    if (!selector) {
      const msg = '`scrollTop` or `selector` is required';
      fail(msg);
      complete(msg);
      return;
    }
    const ele = document.querySelector(selector);
    if (!ele) {
      const msg = `The element cannot be found by the selector \`${selector}\``;
      fail(msg);
      complete(msg);
      return;
    }

    scrollTop = top + ele.getBoundingClientRect().top;
  }

  // Already at the target position
  if (top === scrollTop) {
    success('OK');
    complete('OK');
    return;
  }

  // Scroll immediately
  if (Number(duration) === 0) {
    rootElement.scrollTop = scrollTop;
    success('OK');
    complete('OK');
    return;
  }

  const maxScrollTop = rootElement.scrollHeight - rootElement.clientHeight;
  if (scrollTop < 0) {
    scrollTop = 0;
  } else if (scrollTop > maxScrollTop) {
    scrollTop = maxScrollTop;
  }

  // Scroll distance per 1ms
  const speed = (scrollTop - top) / duration;

  const handleScroll = (startTime: number) => {
    requestAnimationFrame(() => {
      const cost = Date.now() - startTime;
      if (cost < 5) { // throttle
        handleScroll(startTime);
        return;
      }

      top = rootElement.scrollTop;
      let nextTop = top + cost * speed;
      if ((speed > 0 && nextTop > scrollTop) || (speed < 0 && nextTop < scrollTop)) {
        nextTop = scrollTop;
      }

      rootElement.scrollTop = nextTop;
      if (nextTop === scrollTop) {
        success('OK');
        complete('OK');
      } else {
        handleScroll(Date.now());
      }
    });
  };

  // start scrolling
  handleScroll(Date.now());
}, CONTAINER_NAME.WEB);
