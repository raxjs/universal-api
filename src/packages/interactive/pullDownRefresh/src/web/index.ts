import startPullDownRefresh from './startPullDownRefresh';
import stopPullDownRefresh from './stopPullDownRefresh';
import onPullDownRefresh from './onPullDownRefresh';

export {
  onPullDownRefresh,
  startPullDownRefresh,
  stopPullDownRefresh,
};
export default {
  onPullDownRefresh,
  startPullDownRefresh,
  stopPullDownRefresh,
};
