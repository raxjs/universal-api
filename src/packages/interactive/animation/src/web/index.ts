import { Animation, AnimationOptions } from '../types';
import AnimationImpl from './AnimationImpl';

export function createAnimation(options?: AnimationOptions): Animation {
  return new AnimationImpl(options);
}

export default {
  createAnimation,
};
