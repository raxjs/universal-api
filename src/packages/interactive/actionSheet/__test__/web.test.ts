import { sleep, testWebAPI } from '@/utils/__test__/util';
import { DOMUtil } from '@/utils/__test__/web';

testWebAPI('actionSheet', async () => {
  const { showActionSheet } = require('../src/index');
  let res = showActionSheet({
    itemList: ['aaa', 'bbb'],
  });
  expect(document.body).toMatchSnapshot();

  // select item
  new DOMUtil('.__universal_actionsheet_item', 1).trigger('click');
  await expect(res).resolves.toEqual({ tapIndex: 1 });
  expect(document.body).toMatchSnapshot();

  // click cancel
  res = showActionSheet({
    itemList: ['ccc', 'ddd'],
  });
  expect(document.body).toMatchSnapshot();
  new DOMUtil('.__universal_actionsheet_cancel').trigger('click');
  await expect(res).resolves.toEqual({ tapIndex: -1 });
  expect(document.body).toMatchSnapshot();

  // click mask
  res = showActionSheet({
    itemList: ['eee', 'fff'],
  });
  expect(document.body).toMatchSnapshot();
  new DOMUtil('.__universal_actionsheet_mask').trigger('click');
  await expect(res).resolves.toEqual({ tapIndex: -1 });
  expect(document.body).toMatchSnapshot();

  // await remove
  await sleep(300);
});
