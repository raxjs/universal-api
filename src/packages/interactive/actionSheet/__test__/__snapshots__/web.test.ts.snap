// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Platform API: actionSheet Test container: web 1`] = `
<body>
  <style>
    .__universal_actionsheet {  position: fixed;  left: 0;  right: 0;  bottom: 0;  background: #fff;  color: #000;  box-sizing: border-box;  border-radius: 12px 12px 0 0;  transform: translateY(100%);  -webkit-transform: translateY(100%);  z-index: 9999;  outline: none;  text-align: center;  animation-duration: .3s;  animation-fill-mode: both;}.__universal_actionsheet_mask {  content: "";  position: fixed;  left: 0;  right: 0;  bottom: 0;  top: 0;  margin: auto;  display: block;  z-index: 9999;  background: #000;  opacity: 0;  animation-duration: .3s;  animation-fill-mode: both;}.__universal_actionsheet_item {  padding: 16px 0;  border-bottom: 1px solid #eee;}.__universal_actionsheet_cancel {  padding: 16px 0;  border-top: 10px solid #eee;}@keyframes __universal_actionsheet_slideIn {  0% {    transform: translateY(100%);    -webkit-transform: translateY(100%);  }  100% {    transform: translateY(0);    -webkit-transform: translateY(0);  }}@keyframes __universal_actionsheet_slideOut {  0% {    transform: translateY(0);    -webkit-transform: translateY(0);  }  100% {    transform: translateY(100%);    -webkit-transform: translateY(100%);  }}@keyframes __universal_actionsheet_fadeIn {  0% {    opacity: 0;  }  100% {    opacity: .6;  }}@keyframes __universal_actionsheet_fadeOut {  0% {    opacity: .6;  }  100% {    opacity: 0;  }}
  </style>
  <div>
    <div
      class="__universal_actionsheet_mask"
      style="animation-name: __universal_actionsheet_fadeIn;"
    />
    <div
      class="__universal_actionsheet"
      style="animation-name: __universal_actionsheet_slideIn;"
    >
      <div
        class="__universal_actionsheet_item"
        tabindex="0"
      >
        aaa
      </div>
      <div
        class="__universal_actionsheet_item"
        tabindex="1"
      >
        bbb
      </div>
      <div
        class="__universal_actionsheet_cancel"
        tabindex="-1"
      >
        取消
      </div>
    </div>
  </div>
</body>
`;

exports[`Test Platform API: actionSheet Test container: web 2`] = `
<body>
  <style>
    .__universal_actionsheet {  position: fixed;  left: 0;  right: 0;  bottom: 0;  background: #fff;  color: #000;  box-sizing: border-box;  border-radius: 12px 12px 0 0;  transform: translateY(100%);  -webkit-transform: translateY(100%);  z-index: 9999;  outline: none;  text-align: center;  animation-duration: .3s;  animation-fill-mode: both;}.__universal_actionsheet_mask {  content: "";  position: fixed;  left: 0;  right: 0;  bottom: 0;  top: 0;  margin: auto;  display: block;  z-index: 9999;  background: #000;  opacity: 0;  animation-duration: .3s;  animation-fill-mode: both;}.__universal_actionsheet_item {  padding: 16px 0;  border-bottom: 1px solid #eee;}.__universal_actionsheet_cancel {  padding: 16px 0;  border-top: 10px solid #eee;}@keyframes __universal_actionsheet_slideIn {  0% {    transform: translateY(100%);    -webkit-transform: translateY(100%);  }  100% {    transform: translateY(0);    -webkit-transform: translateY(0);  }}@keyframes __universal_actionsheet_slideOut {  0% {    transform: translateY(0);    -webkit-transform: translateY(0);  }  100% {    transform: translateY(100%);    -webkit-transform: translateY(100%);  }}@keyframes __universal_actionsheet_fadeIn {  0% {    opacity: 0;  }  100% {    opacity: .6;  }}@keyframes __universal_actionsheet_fadeOut {  0% {    opacity: .6;  }  100% {    opacity: 0;  }}
  </style>
</body>
`;

exports[`Test Platform API: actionSheet Test container: web 3`] = `
<body>
  <style>
    .__universal_actionsheet {  position: fixed;  left: 0;  right: 0;  bottom: 0;  background: #fff;  color: #000;  box-sizing: border-box;  border-radius: 12px 12px 0 0;  transform: translateY(100%);  -webkit-transform: translateY(100%);  z-index: 9999;  outline: none;  text-align: center;  animation-duration: .3s;  animation-fill-mode: both;}.__universal_actionsheet_mask {  content: "";  position: fixed;  left: 0;  right: 0;  bottom: 0;  top: 0;  margin: auto;  display: block;  z-index: 9999;  background: #000;  opacity: 0;  animation-duration: .3s;  animation-fill-mode: both;}.__universal_actionsheet_item {  padding: 16px 0;  border-bottom: 1px solid #eee;}.__universal_actionsheet_cancel {  padding: 16px 0;  border-top: 10px solid #eee;}@keyframes __universal_actionsheet_slideIn {  0% {    transform: translateY(100%);    -webkit-transform: translateY(100%);  }  100% {    transform: translateY(0);    -webkit-transform: translateY(0);  }}@keyframes __universal_actionsheet_slideOut {  0% {    transform: translateY(0);    -webkit-transform: translateY(0);  }  100% {    transform: translateY(100%);    -webkit-transform: translateY(100%);  }}@keyframes __universal_actionsheet_fadeIn {  0% {    opacity: 0;  }  100% {    opacity: .6;  }}@keyframes __universal_actionsheet_fadeOut {  0% {    opacity: .6;  }  100% {    opacity: 0;  }}
  </style>
  <div>
    <div
      class="__universal_actionsheet_mask"
      style="animation-name: __universal_actionsheet_fadeIn;"
    />
    <div
      class="__universal_actionsheet"
      style="animation-name: __universal_actionsheet_slideIn;"
    >
      <div
        class="__universal_actionsheet_item"
        tabindex="0"
      >
        ccc
      </div>
      <div
        class="__universal_actionsheet_item"
        tabindex="1"
      >
        ddd
      </div>
      <div
        class="__universal_actionsheet_cancel"
        tabindex="-1"
      >
        取消
      </div>
    </div>
  </div>
</body>
`;

exports[`Test Platform API: actionSheet Test container: web 4`] = `
<body>
  <style>
    .__universal_actionsheet {  position: fixed;  left: 0;  right: 0;  bottom: 0;  background: #fff;  color: #000;  box-sizing: border-box;  border-radius: 12px 12px 0 0;  transform: translateY(100%);  -webkit-transform: translateY(100%);  z-index: 9999;  outline: none;  text-align: center;  animation-duration: .3s;  animation-fill-mode: both;}.__universal_actionsheet_mask {  content: "";  position: fixed;  left: 0;  right: 0;  bottom: 0;  top: 0;  margin: auto;  display: block;  z-index: 9999;  background: #000;  opacity: 0;  animation-duration: .3s;  animation-fill-mode: both;}.__universal_actionsheet_item {  padding: 16px 0;  border-bottom: 1px solid #eee;}.__universal_actionsheet_cancel {  padding: 16px 0;  border-top: 10px solid #eee;}@keyframes __universal_actionsheet_slideIn {  0% {    transform: translateY(100%);    -webkit-transform: translateY(100%);  }  100% {    transform: translateY(0);    -webkit-transform: translateY(0);  }}@keyframes __universal_actionsheet_slideOut {  0% {    transform: translateY(0);    -webkit-transform: translateY(0);  }  100% {    transform: translateY(100%);    -webkit-transform: translateY(100%);  }}@keyframes __universal_actionsheet_fadeIn {  0% {    opacity: 0;  }  100% {    opacity: .6;  }}@keyframes __universal_actionsheet_fadeOut {  0% {    opacity: .6;  }  100% {    opacity: 0;  }}
  </style>
</body>
`;

exports[`Test Platform API: actionSheet Test container: web 5`] = `
<body>
  <style>
    .__universal_actionsheet {  position: fixed;  left: 0;  right: 0;  bottom: 0;  background: #fff;  color: #000;  box-sizing: border-box;  border-radius: 12px 12px 0 0;  transform: translateY(100%);  -webkit-transform: translateY(100%);  z-index: 9999;  outline: none;  text-align: center;  animation-duration: .3s;  animation-fill-mode: both;}.__universal_actionsheet_mask {  content: "";  position: fixed;  left: 0;  right: 0;  bottom: 0;  top: 0;  margin: auto;  display: block;  z-index: 9999;  background: #000;  opacity: 0;  animation-duration: .3s;  animation-fill-mode: both;}.__universal_actionsheet_item {  padding: 16px 0;  border-bottom: 1px solid #eee;}.__universal_actionsheet_cancel {  padding: 16px 0;  border-top: 10px solid #eee;}@keyframes __universal_actionsheet_slideIn {  0% {    transform: translateY(100%);    -webkit-transform: translateY(100%);  }  100% {    transform: translateY(0);    -webkit-transform: translateY(0);  }}@keyframes __universal_actionsheet_slideOut {  0% {    transform: translateY(0);    -webkit-transform: translateY(0);  }  100% {    transform: translateY(100%);    -webkit-transform: translateY(100%);  }}@keyframes __universal_actionsheet_fadeIn {  0% {    opacity: 0;  }  100% {    opacity: .6;  }}@keyframes __universal_actionsheet_fadeOut {  0% {    opacity: .6;  }  100% {    opacity: 0;  }}
  </style>
  <div>
    <div
      class="__universal_actionsheet_mask"
      style="animation-name: __universal_actionsheet_fadeIn;"
    />
    <div
      class="__universal_actionsheet"
      style="animation-name: __universal_actionsheet_slideIn;"
    >
      <div
        class="__universal_actionsheet_item"
        tabindex="0"
      >
        eee
      </div>
      <div
        class="__universal_actionsheet_item"
        tabindex="1"
      >
        fff
      </div>
      <div
        class="__universal_actionsheet_cancel"
        tabindex="-1"
      >
        取消
      </div>
    </div>
  </div>
</body>
`;

exports[`Test Platform API: actionSheet Test container: web 6`] = `
<body>
  <style>
    .__universal_actionsheet {  position: fixed;  left: 0;  right: 0;  bottom: 0;  background: #fff;  color: #000;  box-sizing: border-box;  border-radius: 12px 12px 0 0;  transform: translateY(100%);  -webkit-transform: translateY(100%);  z-index: 9999;  outline: none;  text-align: center;  animation-duration: .3s;  animation-fill-mode: both;}.__universal_actionsheet_mask {  content: "";  position: fixed;  left: 0;  right: 0;  bottom: 0;  top: 0;  margin: auto;  display: block;  z-index: 9999;  background: #000;  opacity: 0;  animation-duration: .3s;  animation-fill-mode: both;}.__universal_actionsheet_item {  padding: 16px 0;  border-bottom: 1px solid #eee;}.__universal_actionsheet_cancel {  padding: 16px 0;  border-top: 10px solid #eee;}@keyframes __universal_actionsheet_slideIn {  0% {    transform: translateY(100%);    -webkit-transform: translateY(100%);  }  100% {    transform: translateY(0);    -webkit-transform: translateY(0);  }}@keyframes __universal_actionsheet_slideOut {  0% {    transform: translateY(0);    -webkit-transform: translateY(0);  }  100% {    transform: translateY(100%);    -webkit-transform: translateY(100%);  }}@keyframes __universal_actionsheet_fadeIn {  0% {    opacity: 0;  }  100% {    opacity: .6;  }}@keyframes __universal_actionsheet_fadeOut {  0% {    opacity: .6;  }  100% {    opacity: 0;  }}
  </style>
</body>
`;
